src/sentry/integrations/request_buffer.py

```diff
@@ -0,0 +1,133 @@
+        """
+
+        return json.loads(redis_object)
+
+    def _get_all_from_buffer(self, buffer_key):
+        """
+        Get the list at the buffer key.
+        """
+
+        return self.client.lrange(buffer_key, 0, BUFFER_SIZE - 1)
+
+    def _get(self):
+        """
+        Returns the list of daily aggregate error counts.
+        """
+        return [
+            self._convert_obj_to_dict(obj) for obj in self._get_all_from_buffer(self.integrationkey)
+        ]
+
+    def is_integration_broken(self):
+        """
+        Integration is broken if we have 7 consecutive days of errors and no successes OR have a fatal error
+
+        """
+        data = [
+            datetime.strptime(item.get("date"), "%Y-%m-%d").date()
+            for item in self._get()
+            if item.get("fatal_count", 0) > 0 and item.get("date")
+        ][0:IS_BROKEN_RANGE]
+
+        if len(data) > 0:
+            return True
+
+        data = [
+            datetime.strptime(item.get("date"), "%Y-%m-%d").date()
+            for item in self._get()
+            if item.get("error_count", 0) > 0
+            and item.get("success_count", 0) == 0
+            and item.get("date")
+        ][0:IS_BROKEN_RANGE]
+
+        if not len(data):
+            return False
+
+        if len(data) < IS_BROKEN_RANGE:
+            return False
+
+        date_set = {data[0] - timedelta(x) for x in range((data[0] - data[-1]).days)}
+        missing = list(date_set - set(data))
+        if len(missing):
+            return False
+        return True
+
+    def add(self, count: str):
+        VALID_KEYS = ["success", "error", "fatal"]
+        if count not in VALID_KEYS:
+            raise Exception("Requires a valid key param.")
+
+        other_count1, other_count2 = list(set(VALID_KEYS).difference([count]))[0:2]
+        buffer_key = self.integrationkey
+        now = datetime.now().strftime("%Y-%m-%d")
+
+        pipe = self.client.pipeline()
+
+        # get first element from array
+        recent_item_array = self.client.lrange(buffer_key, 0, 1)
+        if len(recent_item_array):
+            recent_item = json.loads(recent_item_array[0])
+            if recent_item.get("date") == now:
+                recent_item[f"{count}_count"] += 1
+                pipe.lset(buffer_key, 0, json.dumps(recent_item))
+            else:
+                data = {
+                    "date": now,
+                    f"{count}_count": 1,
+                    f"{other_count1}_count": 0,
+                    f"{other_count2}_count": 0,
+                }
+                pipe.lpush(buffer_key, json.dumps(data))
+
+        else:
+            data = {
+                "date": now,
+                f"{count}_count": 1,
+                f"{other_count1}_count": 0,
+                f"{other_count2}_count": 0,
+            }
+            pipe.lpush(buffer_key, json.dumps(data))
+
+        pipe.expire(buffer_key, KEY_EXPIRY)
```

Comment:
if add keeps being called, the key will never expire and the list will only expand?

Reply:
Is it being done wrong? I'm not super familiar with this but the [docs](https://redis.io/commands/expire/) here show you can set an expiration date, so isn't this function just adding to the list and setting an expiration date?

Reply:
A list-type key in redis is still a single key, with a single ttl. assuming add is always called within KEY_EXPIRY seconds, the list (and all of it's values) will never expire since the expiration would keep getting reset to the future.

Reply:
We changed the `buffer_key` to `self.integrationkey + now` so `pipe.expire(buffer_key, KEY_EXPIRY)` will be called with different key for each day giving the key time to expire. However, `test_expiry` in` test_disable.py` is still showing that are keys are not expiring. Would appreciate any suggestions!

Reply:
I'm not aware of the exact testing setup, but deletion of expired keys in redis is not guaranteed to happen immediately after the expiration passes.
> In Redis 2.4 the expire might not be pin-point accurate, and it could be between zero to one seconds out.
>
> Since Redis 2.6 the expire error is from 0 to 1 milliseconds.

Reply:
It looks like expiring elements within a redis list is a common feature request but something that isn't implemented. A suggested workaround offered by the developers of redis would be to use a sorted set: https://github.com/redis/redis/issues/135

Reply:
We decided on using `ltrim` which should work for our purposes 🙃

Reply:
Good solution with ltrim, since the list is already sorted due to the way it's written. One thing to note is that this might have a race condition between calling LRANGE on buffer_key and executing the pipeline. You may need to use WATCH as described [here](https://redis-py.readthedocs.io/en/stable/advanced_features.html#default-pipelines).

Reply:
To expand on this a bit, the redis data structure you could use is a sorted set where key is unchanged with what you already have (`self.integration_key`) and value is a sorted set where score is the day (you'll need to convert this to an int `int((datetime.now() + timedelta(days=expiration_days)).replace(hour=0,minute=0,microsecond=0).timestamp())` when you logged the request and the value is a json blob of counts.

To modify your `add` method, you'll want to: 
1. do a `ZRANGEBYSCORE <buffer_key> <now-datetime-int> +inf LIMIT 1` to replace your `lrange(buffer_key, 0, 1)` to get the most recent day count. 
2. apply a `ZREMRANGEBYSCORE <buffer_key> -inf <now + KEY_EXPIRATION>)` to explicitly expire out any day aggregated counts that are older than KEY_EXPIRATION.
3. continue setting the expiration for the entire sorted set `pipe.expire(buffer_key, KEY_EXPIRY)` just so inactive sorted sets will get cleaned up.

Reply:
Looks like `ZRANGEBYSCORE` is deprecated - https://redis.io/commands/zrangebyscore/ but it gives an alternative

Reply:
@ellisonmarks Thanks for the heads up! Implemented WATCH

Reply:
Python docs for `zrangebyscore`: https://redis-py.readthedocs.io/en/stable/commands.html#redis.commands.core.CoreCommands.zrangebyscore

Tried to get this working but kinda ran out of time, I'll try to help tomorrow and then we can add the `WATCH` stuff.